include/deal.II/lac/trilinos_sparse_matrix.h

```diff
@@ -2686,9 +3098,232 @@ namespace TrilinosWrappers
 
 
 
+  namespace internal
+  {
+    namespace LinearOperator
+    {
+
+      template <typename Solver, typename Preconditioner>
+      TrilinosPayload
+      inverse_payload (const TrilinosPayload &payload,
+                       Solver                &solver,
+                       const Preconditioner  &preconditioner)
+      {
+        TrilinosPayload return_op(payload);
+
+        // Capture by copy so the payloads are always valid
+
+        return_op.inv_vmult = [payload,&solver,&preconditioner](
+                                TrilinosPayload::Domain &tril_dst, const TrilinosPayload::Range &tril_src
+                              )
+        {
+          // Duplicated from TrilinosWrappers::PreconditionBase::vmult
+          // as well as from TrilinosWrappers::SparseMatrix::Tvmult
+          Assert (&tril_src != &tril_dst, TrilinosWrappers::SparseMatrix::ExcSourceEqualsDestination());
+          internal::check_vector_map_equality(payload,
+                                              tril_src, tril_dst,
+                                              !payload.UseTranspose());
+          solver.solve(payload, tril_dst, tril_src, preconditioner);
+        };
+
+        return_op.inv_Tvmult = [payload,&solver,&preconditioner](
+                                 TrilinosPayload::Range &tril_dst, const TrilinosPayload::Domain &tril_src
+                               )
+        {
+          // Duplicated from TrilinosWrappers::PreconditionBase::vmult
+          // as well as from TrilinosWrappers::SparseMatrix::Tvmult
+          Assert (&tril_src != &tril_dst, TrilinosWrappers::SparseMatrix::ExcSourceEqualsDestination());
+          internal::check_vector_map_equality(payload,
+                                              tril_src, tril_dst,
+                                              payload.UseTranspose());
+
+          const_cast<TrilinosPayload &>(payload).transpose();
+          solver.solve(payload, tril_dst, tril_src, preconditioner);
+          const_cast<TrilinosPayload &>(payload).transpose();
+        };
+
+        // If the input operator is already setup for transpose operations, then
+        // we must do similar with its inverse.
+        if (return_op.UseTranspose() == true)
+          std::swap(return_op.inv_vmult,
+                    return_op.inv_Tvmult);
+
+        return return_op;
+      }
+
+
+      inline IndexSet
+      TrilinosPayload::locally_owned_domain_indices () const
+      {
+        return IndexSet(domain_map);
```

Comment:
This and the following few functions all look pretty simple but are ultimately not trivial in what they do. There is probably little need to make them `inline`.